examples/adaptive_crag/application.py

```diff
@@ -0,0 +1,537 @@
+from pathlib import Path
+from typing import Iterable, Literal
+
+import google.generativeai as genai
+import instructor
+import lancedb
+from burr.core import Application, ApplicationBuilder, State, expr, when
+from burr.core.action import action
+from burr.tracking import LocalTrackingClient
+from dotenv import load_dotenv
+from exa_py.api import Exa
+from lancedb.db import DBConnection
+from lancedb.embeddings import get_registry
+from lancedb.pydantic import LanceModel, Vector
+from lancedb.table import Table
+from pydantic import BaseModel, Field
+
+load_dotenv()  # load your GOOGLE_API_KEY and EXA_API_KEY from a .env file
+
+
+ask_gemini = instructor.from_gemini(
+    client=genai.GenerativeModel(model_name="gemini-1.5-flash-latest")
+)
+exa = Exa(api_key=None)
+
+BURR_DOCS_DIR = Path("burr_docs")
+DOC_THRESH = 512
+LANCE_URI = "lance/rag"
+EMS_MODEL = "BAAI/bge-small-en-v1.5"  # for the embeddings in LanceDB
+DEVICE = "cuda"  # or "cpu"
+DOCS_LIMIT = 5  # number of documents to retrieve from the database or search engine
+ATTEMPTS = 3  # ask_gemini will try to call `create` this many times before giving up
+
+
+def chat_message(role: str, content: str) -> dict[str, str]:
```

Comment:
Thoughts on starting these helper functions with `_` to visually differentiate them from the Burr functions?

Reply:
Is the @action decorator not enough of a differentiation?

Reply:
Normally I'd say it is but for an example I thnk it might be nice to have it as `_` to bring attention to the `@action` ones. Up to you though.